### ISSUE TITLE
Different Public key issue when Signing Ethereum Raw Transaction

### ISSUE THREAD

**Issue by Hoodie-Kang** [getamis/alice#236]
Hi

I'm trying to send a raw transaction in the Goerli Ethereum test net using Web3.js.
Here is the transaction data and the hash of the transaction.

```
{
  nonce: '0x0',
  to: '0x66Cba25faF0a341Fe536925c46687B1160a197a7',
  value: '0x38d7ea4c68000',
  gasLimit: '0x1d8a8',
  gasPrice: '0x4a817c800',
  data: '0x0',
  v: '0x2d',
  r: '0x0',
  s: '0x0'
}
hash: 3a6e0c03fcfd2ac737a2fa8a9535a088469e1f75f323a8377f8355a4103ba00f
```

Using GG18, after 2 of 2 DKG the signer inputs were:
```
port: 10001
peers:
  - 10002
share: "48009635119716801121422388540381921473982761023954328247557439033967871640185"
pubkey:
  x: "10290526619007406802679274811427206866549400507245539291039072171369241278468"
  y: "37504811230942583636922232634339105768466278733205375082142890370196805216860"
bks:
  id-10001:
    x: "9987879638869391180574507371944186878398536146226961513235596770885926708560"
    rank: 0
  id-10002:
    x: "27430016224955743730177942842935577266812994817782174086985731142993166568517"
    rank: 0
msg: "3a6e0c03fcfd2ac737a2fa8a9535a088469e1f75f323a8377f8355a4103ba00f"
```
```
port: 10002
peers:
  - 10001
share: "83571483614016019445058258921935883022592627804235010868567212221454054200457"
pubkey: same
bks: same
msg: same
```

the sign output
```
r: "89604550539853586697799468095878687627444370844130366237671645935096428765871"
s: "102093535849068150100945179871541286901271488417765717028400532673319260928573"
```

I put it in the transaction data and send the transaction.

And here is the problem..

I guess the address of the sender is "0x8B217caa3B2086C8C3769621D9754c11C59b998D"
(ref. https://ethereum.org/en/developers/docs/accounts/ ) 

But when I derive the sender address by using getSenderAddress() in 'ethereumjs-tx', the sender addresses are not same.

I'm very confused :(
Can you help me to find out what the problem is?

Thank you always,
Hoodie

p.s.   Even using CGGMP, above problem also happened ..
p.s.2 I edit msg code as msg, _ := hex.DecodeString(config.Message) because hash of transaction is hex string format

**Comment by cychuang0924** [comment]
I think this issue which can help you! https://github.com/getamis/alice/issues/117
If you have any problems, please let me know. Thanks a lot!

**Comment by Hoodie-Kang** [comment]
Hi Chihyun Chuang,

Actually I already read the issue and the comment , and edited some codes dealing with sign result to expose V bit.

And the sign works but the problem is that, as I said, both case of V bit [0/1] the TWO addresses are different.

So I think it's another kind of issue.

Is there any other possibilities? Thanks!

**Comment by cychuang0924** [comment]
Hi Hoodie-Kang,
     My colleague @alanchchen  uses the public key as follows:
     ```
x, _ := new(big.Int).SetString("10290526619007406802679274811427206866549400507245539291039072171369241278468", 10)
	y, _ := new(big.Int).SetString("37504811230942583636922232634339105768466278733205375082142890370196805216860", 10)
	p, err := ecpointgrouplaw.NewECPoint(
		elliptic.Secp256k1(),
		x,
		y,
	)
	if err != nil {
		panic(err)
	}

	pubKey := p.ToPubKey()
	fmt.Println("address", crypto.PubkeyToAddress(*pubKey))
	pubKeyByte := crypto.FromECDSAPub(pubKey)
	fmt.Println("pubkey", hex.EncodeToString(pubKeyByte))
```
to obtain your address 
`address 0x8555Df2fc424Af3eD10B4D94E0258Ef03CCD5308
pubkey 0416c03b4e4a545146e5154d2b6415f5c23dbfa65c327b67f8cc2e783dadb0780452eaf88cfb5927067328fcc22454a5dcfd82728fcea3641d94e2a0940ae36a5c`


If it is correct, then your message maybe be a wrong format. One possibility is that your can not use []byte("3a6e0c03fcfd2ac737a2fa8a9535a088469e1f75f323a8377f8355a4103ba00f") to transfer it.

You should use  some tools to transform hex to byte array.

If you have more information, please offer us. Thanks a lot!

Sincerely,

Chih Yun

**Comment by Hoodie-Kang** [comment]
Hi Chih Yun,
Thank you for your helpful comment.

First, my address from above calculations and your colleagues' address are different. And I think maybe his address is correct. Thanks! I will edit my code.

Second, you mentioned that _your message maybe be a wrong format._
I also thought about the possibility so I added some code like below at example/signer/service.go:58
```
	fmt.Println(hex.DecodeString(config.Message))
	// Create signer
	msg, _ := hex.DecodeString(config.Message)
	signer, err := signer.NewSigner(pm, dkgResult.PublicKey, paillier, dkgResult.Share, dkgResult.Bks, msg, s)
```
and I put msg as I mentioned above comment,
```
msg: "3a6e0c03fcfd2ac737a2fa8a9535a088469e1f75f323a8377f8355a4103ba00f"
```

In short, the problem is that the sender address from the signature changes every time I make a new signature of the same msg.

For more information, I put the whole code I've been using to send raw transaction in the Goerli testnet.

```
const ethTx = require('ethereumjs-tx').Transaction;
const Web3 = require('web3');

let web3;
let received;

async function main() {
    received = '0x66Cba25faF0a341Fe536925c46687B1160a197a7'; // my metamask account
    var provider = 'https://goerli.infura.io/v3/3b5c9fc2f01a4c4c982f4b81d9b841ab'
    web3 = new Web3(new Web3.providers.HttpProvider(provider));

    // my calculation to create address from ECDSA public key
    // TODO: maybe wrong. should fix it
    // const x = web3.utils.toHex("10290526619007406802679274811427206866549400507245539291039072171369241278468")
    // const y = web3.utils.toHex("37504811230942583636922232634339105768466278733205375082142890370196805216860")
    // const pub = x.slice(2)+ y.slice(2)
    // const pubkey = web3.utils.sha3(pub).slice(-40)
    // console.log("0x"+pubkey)
    // console.log(web3.utils.sha3(pubkey).slice(0,42))

    // creating tx
    var txData = {
        nonce: '0x0', //web3.utils.toHex(await web3.eth.getTransactionCount(test_sender)),
        to: received,
        value: web3.utils.toHex(web3.utils.toWei('0.001', 'ether')),
        gasLimit: web3.utils.toHex(121000),
        gasPrice: web3.utils.toHex(web3.utils.toWei('20', 'gwei')),
        data: web3.utils.toHex(''),
        v: '0x2d',
        r: '0x0',
        s: '0x0'
    };
    console.log(txData)
    var tx = new ethTx(txData, {'chain':'goerli'});
    txHash = tx.hash(true);

    // msg
    console.log(txHash.toString('hex'));

    // after sign
    tx.v = '0x2e'
    r = web3.utils.toHex('60338192179111296923593312738172596742629872261009372225656975881647879931059');
    s = web3.utils.toHex('24770887441067885354558877242527325168918232888349381462116062864195195367716');
    tx.r = r
    tx.s = s
    // get sender address
    console.log(tx.getSenderAddress().toString('hex'))
    console.log(tx.getSenderPublicKey())

    // send transaction
    // If the address I guess and the sender address from the code are different,
    // then Error: Returned error: insufficient funds for gas * price + value
    // Because I don't have any eth in the sender address(It's not what I guessed!)
    const serializedTx = tx.serialize();
    const receipt = await web3.eth
        .sendSignedTransaction('0x' + serializedTx.toString('hex'))
        .once("transactionHash", (txhash) => {
            console.log(`mining transaction ...`);
            console.log(`https://goerli.etherscan.io/tx/${txhash}`);
        });
    // the tx is now on chain!
    console.log(`Mined in block ${receipt.blockNumber}`);
}

main();

```

**Comment by cychuang0924** [comment]
Hi Hoodie Kang,

       Let me first explain this phenomenon：**In short, the problem is that the sender address from the signature changes every time I make a new signature of the same msg.**
       
       At first, the address is derived from the public key (i.e. use SHA3). Second, although signatures produced by the same private key and message will change every time, you still can use the same public key to verify the correctness of these 
signatures. That means **you can recover the same public key from your signatures and the message.** The explanation can be found in [here](https://medium.com/asecuritysite-when-bob-met-alice/crypto-magic-recovering-alices-public-key-from-an-ecdsa-signature-e7193df8df6e). 

Combing these two facts, you should get the same address in your case.

So far, we have not solved your question, but we pointed this strange  phenomenon.

Thanks for your information. We will keep on figuring out it.

**Comment by Hoodie-Kang** [comment]
Hi Chih Yun,

Thanks for your explanation, I understand the computation process.
And I'll also try hard to find out the answer!

Thank you,
Hoodie Kang

**Comment by Hoodie-Kang** [comment]
Hi all!

Finally I sent a raw transaction successfully and that means sender address "is" the address that I made with the public key!!

Here is my solution. There were two kinds of problems that I fixed:
First, in my code
```
    txHash = tx.hash(true)
->txHash = tx.hash(false)
```
The boolean argument means "isIncludeSignature" so I changed the value to "false".
Then it changed the hash msg and I used it to sign.

Second, the exposing V bit code that you mentioned above helped.
```
    S <= Secp256k1.N / 2
```
The S should not be bigger than N of Secp256k1 divided by 2. (ref. https://eips.ethereum.org/EIPS/eip-2)
So I made a signature only when the S is less than or equal to N/2. And then It worked with both of CGGMP and GG18.

Thank you so much for your hard work to find a solution to this issue.

Sincerely,
Hoodie Kang
